### PR TITLE
python: implement IoT3 Core SDK

### PR DESCRIPTION
**Pull request by ymorin-orange**
---
**Features**;

* Implement IoT3 Core SDK:
  * MQTT client
  * OpenTelemetry client
  * telemetry sent for MQTT messages

Closes: #123
Closes: #122

---
**How to test**

1. Prepare an OpenTelemetry collector on localhost:
    ```
    $ docker container run \
        --rm \
        -p 16686:16686 \
        -p 4318:4318 \
        jaegertracing/all-in-one:1.58
    ```
    then open a browser on the Jaegger UI:
    ```
    http://localhost:16686/
    ```
2. Prepare an MQTT broker on localhost (adapt based on your distribution):
    ```sh
    $ sudo systemctl start mosquitto.service
    ```
    and then start an MQTT client to dump the messages on the broker:
    ```
    $ mosquitto_sub -V 5 -t '#' -F '%j' |jq .
    ```
3. Test IoT3 Core SDK:
    1. start an environment with python 3.11:
        ```sh
        $ docker container run \
            --rm \
            -ti \
            --network host \
            -e http_proxy \
            -e https_proxy \
            -e no_proxy \
            --user $(id -u):$(id -u) \
            --mount type=bind,source=$(pwd),destination=$(pwd) \
            --workdir $(pwd) \
            --entrypoint /bin/bash \
            python:3.11.9-slim-bookworm
        ```
    2. create a /venv/ in a writable location, and activate it:
        ```sh
        (docker) python3 -m venv /tmp/iot3
        (docker) . /tmp/iot3/bin/activate
        ```
    3. install the IoT3 Core SDK (notice the trailing dot `.`):
        ```sh
        (docker) pip3 --disable-pip-version-check install .
        ```
    4. Run the MQTT-only test:
        ```sh
        (docker) ./tests/test-iot3-core-mqtt
        ```
    5. Run the OpenTelemetry-only test:
        ```sh
        (docker) ./tests/test-iot3-core-otel
        ```
        then, in Jaegger, search for the traces for the `test-service`
        service.
    6. Run the MQTT with OpenTelemetry test:
        ```sh
        (docker) ./tests/test-iot3-core-all
        ```
        then, in Jaegger, search for the traces for the `test-service`
        service.
    7. Run the simple API test:
        ```sh
        (docker) ./tests/test-iot3-core
        ```
        then, in Jaegger, search for the traces for the `my-service`
        service.

---
**Expected results:**

1. The Jaegger UI is displayed in your browser
2. The mosquitto server is running
3. The IoT3 Core SDK works;
    1. the docker container runs
    2. the python /venv/ is created and activated
    3. the IoT3 Core SDK is installed
    4. the MQTT message is dumped by /mosquitto_sub/:
        ```json
        {
          "tst": "2024-07-31T13:27:08.106723Z+0200",
          "topic": "foo/bar",
          "qos": 0,
          "retain": 0,
          "payloadlen": 5,
          "payload": "pouet"
        }
        ```
    5. the Jaegger UI finds two traces:
        * one with 6 spans, of which the last one is in error, reporting /Some failure/
        * one with a single span, which has a link to the root span
          of the first trace, above
    6. the MQTT message is dumped by /mosquitto_sub/:
        ```json
        {
          "tst": "2024-07-31T13:42:33.339973Z+0200",
          "topic": "foo/bar/passed",
          "qos": 0,
          "retain": 0,
          "payloadlen": 6,
          "properties": {
            "user-properties": {
              "traceparent": "00-cb2e0b60496eca82dfb9420006759aff-ec4169b906b4a5cd-00"
            }
          },
          "payload": "passed"
        }
        ```
        then Jaegger UI finds three new traces, all with a single span:
        * the first is for the failed publish, and has a span:
            * of kind `producer`
            * with attributes:
                * `test.iot3.core.mqtt.action=publish`
                * `test.iot3.core.mqtt.topic=foo/bar/dropped`
            * with an error status reporting /The client is not currently connected./
        * the second is for the sucessful publish, and has a span:
            * of kind `producer`
            * with attributes:
                * `test.iot3.core.mqtt.action=publish`
                * `test.iot3.core.mqtt.topic=foo/bar/passed`
            * with an unset status
        * the third is for the receive, and has a span:
            * of kind `consumer`
            * with attributes:
                * `test.iot3.core.mqtt.action=receive`
                * `test.iot3.core.mqtt.topic=foo/bar/passed`
            * with an unset status
            * with a link to the span of the second trace, above
    7. the results are similar to those of step 6, above

### CODE REVIEW

**Comment by ymorin-orange** [comment]
* rebased on top of `master` after the merge of #147
* really use `zlib.compress()` instead of `gzip.compress()`

**Comment by ymorin-orange** [comment]
See https://github.com/ymorin-orange/its-client/tree/yem/its-iot3 for examples on how it can get used in our own applications (to be merged later).

**Comment by ymorin-orange** [comment]
Updated to account for _black_ reformatting.

Bizarrely enough, _black_ in the workflows did not catrch the issues, but it was caught locally... :-/

**Comment by tigroo** [comment]
tests ok:

```
(kvmk8371) kvmk8371@yd-89lzhs3:~/Workspace/its-client$ mosquitto_sub -V 5 -t '#' -F '%j' |jq .
{
  "tst": "2024-08-30T18:45:10.699818Z+0200",
  "topic": "foo/bar",
  "qos": 0,
  "retain": 0,
  "payloadlen": 5,
  "payload": "pouet"
}
{
  "tst": "2024-08-30T18:52:23.298143Z+0200",
  "topic": "foo/bar/passed",
  "qos": 0,
  "retain": 0,
  "payloadlen": 6,
  "properties": {
    "user-properties": {
      "traceparent": "00-7c3105bbd6f4b319c26da241e575359e-44e7dd9181bb0566-00"
    }
  },
  "payload": "passed"
}

```

Captura d'ecran de 2024-08-30 18-49-41
Captura d'ecran de 2024-08-30 18-49-50
Captura d'ecran de 2024-08-30 18-51-44
Captura d'ecran de 2024-08-30 18-55-06

File: python/iot3/src/iot3/core/__init__.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+from . import otel as _otel
+
+"""
+This module provides a simple API wrapping MQTT and OpenTelemetry.
+
+The simple API consists in a few functions and a type annotation:
+
+* Functions:
+  * start(): start the MQTT and the OpenTelemetry clients
+  * ready(): returns whether the MQTT client is ready
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:21]: return

File: python/iot3/src/iot3/core/__init__.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    message_callback: _Optional[MsgCallbackType] = None,
+    callback_data: _Any = None,
+) -> None:
+    """Start the IoT3 Core SDK.
+
+    :param config: The SDK configuration.
+    :param message_callback: The function to call when a message is
+                             received; when called, the first argument
+                             will be callback_data, the second will be
+                             the topic the emsage was received on, and
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:78]: message

File: python/iot3/src/iot3/core/__init__.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    if _core is None:
+        raise RuntimeError("IoT3 Core SDK not initialised.")
+
+    _core["mqtt"].stop()
+    _core["otel"].stop()
+    _core = None
+
+
+def is_ready() -> bool:
+    """Checkxs whether the IoT3 Core SDK is ready.
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:149]: Checks

File: python/iot3/src/iot3/core/__init__.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    if _core is None:
+        raise RuntimeError("IoT3 Core SDK not initialised.")
+
+    return _core["mqtt"].is_ready()
+
+
+def wait_for_ready():
+    """Wait until the IoT2 Core SDK is ready.
+
+    Beware that this may take an undeterminate amount of time; in
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:166]: undetermined

File: python/iot3/src/iot3/core/__init__.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    case the MQTT client can't connect at all, wait_for_ready()
+    will wait forever and never return.
+    """
+    while True:
+        if ready():
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:171]: is_ready()?

File: python/iot3/src/iot3/core/__init__.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    global _core
+
+    if _core is None:
+        raise RuntimeError("IoT3 Core SDK not initialised.")
+
+    topics = topics if isinstance(topics, list) else [topics]
+    _core["mqtt"].subscribe_replace(topics=topics)
+
+
+_core = None
```
Comment (tigroo) [inline python/iot3/src/iot3/core/__init__.py:211]: Could you explain me?

File: python/iot3/src/iot3/core/mqtt.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+            host=self.host,
+            port=self.port,
+            clean_start=True,
+        )
+        self.client.loop_start()
+
+    def is_ready(self):
+        """Returns whether the MQTT client is ready.
+
+        The client is ready when it is connected to the MQTT borker.
```
Comment (tigroo) [inline python/iot3/src/iot3/core/mqtt.py:93]: broker

File: python/iot3/src/iot3/core/mqtt.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        span_ctxmgr_cb: Optional[SpanCallableType] = otel.Otel.noexport_span,
+    ):
+        """
+        Create an MQTT client
+
+        :param client_id: The MQT client ID to identify as.
+        :param host: The host name (or IP) of the MQTT broker.
+        :param port: The port the MQTT broker listens on.
+        :param socket_path: The path of the UNIX socket.
+        :param username: The username to authenticate agaisnt the MQTT broker.
```
Comment (tigroo) [inline python/iot3/src/iot3/core/mqtt.py:41]: against

File: python/iot3/src/iot3/core/mqtt.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        :param password: The password to authenticate agaisnt the MQTT broker.
```
Comment (tigroo) [inline python/iot3/src/iot3/core/mqtt.py:42]: against

File: python/iot3/src/iot3/core/otel.py [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+
+    def to_traceparent(self):
+        """Return a traceparent that refers to the span."""
+        return f"00-{self.trace_id}-{self.span_id}-00"
+
+
+class NoExportSpan(Span):
+    """A fake span when no telemetry is used"""
+
+    def to_dict():
```
Comment (tigroo) [inline python/iot3/src/iot3/core/otel.py:490]: def to_dict(self): ?

File: python/iot3/tests/test-iot3-core-mqtt [hunk 11]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python3
+
+import iot3.core.mqtt
+import time
+
+def recv(data, topic, payload):
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core-mqtt:6]: PEP 8: E302 expected 2 blank lines, found 1

File: python/iot3/tests/test-iot3-core-mqtt [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    print(f"{topic[:16]}: {payload[:16]}")
+
+c = iot3.core.mqtt.MqttClient(
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core-mqtt:9]: PEP 8: E305 expected 2 blank lines after class or function definition, found 1

**Review by tigroo — CHANGES_REQUESTED** [review]

File: python/iot3/src/iot3/core/__init__.py (hunk 4 above)
Comment (ymorin-orange) [reply python/iot3/src/iot3/core/__init__.py:166]: There was indeed a typo. It's in fact spelled [_indeterminate_](https://www.merriam-webster.com/dictionary/indeterminate). :-)

File: python/iot3/src/iot3/core/__init__.py (hunk 6 above)
Comment (ymorin-orange) [reply python/iot3/src/iot3/core/__init__.py:211]: This variable holds the context of the simple API. It is accessed as
a `global` in the various functions, so it must be defined at the
module level.

**Review by tigroo — APPROVED** [review]
code ok, tests ok

**Pull request merged**